### PR TITLE
liquify handles

### PR DESCRIPTION
**Pull request by TurboGit**
better control over the liquify nodes at any resolution/zoom.

### CODE REVIEW

**Comment by TurboGit** [comment]
Please, test this last version makes the liquify controls far more fluid to use.

File: src/iop/liquify.c [hunk 1]
```diff
@@ -2268,9 +2183,219 @@ static float find_nearest_on_curve_t(const float complex p0,
...
+          // sides.
+          const float complex deadzone = (point - prev->warp.point) / 20.0f;
+          const float complex lp1 = prev->warp.point + deadzone;
+          const float complex lp2 = point - deadzone;
+          const float t = find_nearest_on_line_t(lp1, lp2, *pt);
+
+          if(t > 0.0f && t < 1.0f)
+          {
+            const float complex linepoint = cmix(lp1, lp2, t);
+            const float d = cabsf(linepoint - *pt);
```
Comment (aurelienpierre) [inline src/iop/liquify.c:2268]: d should be a euclidean distance, no ? aka hypotf(x_cursor - x_node, y_cursor - y_node).

Comment (TurboGit) [reply src/iop/liquify.c:2268]: if I'm not mistaken this is what cabsf() does here.

Comment (da-phil) [reply src/iop/liquify.c:2268]: The `cabs(z)` function is equivalent to `hypot(z.x, z.y)`, so you're both right :smile:

**Comment by rawfiner** [comment]
Tested quickly, much better indeed! Thanks!

**Comment by TurboGit** [comment]
Thanks @rawfiner for testing. Indeed much better too. I suppose this will help even more on slow computer has now moving the forms is really light on the CPU.

File: src/iop/liquify.c (hunk 1 above)
Comment (aurelienpierre) [reply src/iop/liquify.c:2268]: I'm just concerned that dealing with complex numbers will be slower than dealing with a vector of floats

**Comment by aurelienpierre** [comment]
Code looks good, feature works much better, thanks !

2 minor comments, though :
1. to drag the vector, one has to drag the point of the arrow, while we would expect to interact with the center of gravity of the arrow
2. on HighDPI (×2) screens, you need to be super accurate on node selection. Maybe making the radius of sensivitity proportional to DPI factor (`DT_PIXEL_APPLY_DPI()` macro) should help.

**Comment by TurboGit** [comment]
>     1. to drag the vector, one has to drag the pointy head of the arrow, while we would expect to interact with the center of gravity of the arrow,

fixed.

>     2\. on HighDPI (×2) screens, you need to be super accurate on node selection. Maybe making the radius of sensivitity proportional to DPI factor (`DT_PIXEL_APPLY_DPI()` macro) should help.

Fixed too.

Can you double check and report ?

File: src/iop/liquify.c (hunk 1 above)
Comment (TurboGit) [reply src/iop/liquify.c:2268]: No, behind the scene this is just a set of two float. Nothing more, nothing less and all nice routines coming with them are very useful in a two space canvas.

**Comment by aurelienpierre** [comment]
It works beautifully, thanks a lot !

**Comment by TurboGit** [comment]
Merged in master and queued for 3.4.1.

**Pull request merged**